File: 2019/golang/day01/fuelcounter.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type Module struct {
	mass int
}

func main() {
	modules := Modules("../../puzzledata/day01/input.txt")
	fmt.Println("Puzzle One Solution:", PuzzleOne(modules))
	fmt.Println("Puzzle Two Solution:", PuzzleTwo(modules))
}

func PuzzleOne(modules []Module) int {
	var sum int
	for _, module := range modules {
		sum += CalculateFuel(module.mass)
	}
	return sum
}

func PuzzleTwo(modules []Module) int {
	var sum int
	for _, module := range modules {
		sum += CalculateFuelRecursive(module.mass, 0)
	}
	return sum
}

func Modules(datapath string) []Module {
	file, err := os.Open(datapath)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	var modules []Module
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		line = strings.TrimSuffix(line, "\n")

		mass, err := strconv.Atoi(line)
		if err != nil {
			log.Fatal(err)
		}
		modules = append(modules, Module{mass})
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	return modules
}

func CalculateFuel(mass int) int {
	return mass/3 - 2
}

func CalculateFuelRecursive(mass int, fuelSum int) int {
	fuel := CalculateFuel(mass)
	if fuel <= 0 {
		return fuelSum
	}
	return CalculateFuelRecursive(fuel, fuel+fuelSum)
}

File: 2019/python3/day01/fuel_counter.py
import os

def main():
    total_fuel_requirement = 0
    for mass in read_input():
        total_fuel_requirement += calculate_fuel_requirement(mass)
    print(f"Total Fuel Requirement: {total_fuel_requirement}")


def read_input():
    curdir = os.path.dirname(__file__) 
    datapath = os.path.join(curdir, './../../puzzledata/day01/input.txt')

    with open(datapath, 'r') as file:
        for line in file:
            yield int(line.strip())


def calculate_fuel_requirement(mass, total_fuel=0):
    fuel = mass // 3 - 2
    if fuel <= 0:
        return total_fuel
    return calculate_fuel_requirement(fuel, total_fuel+fuel) 


if __name__ == "__main__":
    main()